extensions/amp-ad-network-base/0.1/amp-ad-network-base.js

```diff
@@ -0,0 +1,287 @@
+   *    current ad retrieval is current (meaning unlayoutCallback was not
+   *    executed). If not, will return false.
+   * @throws {Error}
+   */
+  getFreshnessVerifier() {
+    const id = this.freshnessId_;
+    return () => {
+      if (id != this.freshnessId_) {
+        this.handleStaleExecution();
+        return true;
+      }
+      return false;
+    };
+  }
+
+  /**
+   * @param {!ValidatorOutputDef} validatorOutput
+   * @return {!RendererInputDef}
+   */
+  getRenderingDataInput_(validatorOutput) {
+    const creative = validatorOutput.creative;
+    return /** @type {!RendererInputDef} */ ({
+      creativeMetadata: getAmpAdMetadata(creative, TAG, this.networkType_),
+      templateData: null,
+      crossDomainData: {
+        rawCreativeBytes: this.unvalidatedBytes_,
+        additionalContextMetadata: this.getCrossDomainContextData(),
+        sentinel: this.getCrossDomainSentinel(),
+      },
+      unvalidatedBytes: this.unvalidatedBytes_,
+      // TODO(levitzky) This may change based on the ad response.
+      size: this.initialSize_,
+      adUrl: this.expandedAdUrl_ ? this.expandedAdUrl_ : this.adUrl_,
+    });
+  }
+
+  /**
+   * Processes the ad response as soon as the XHR request returns. This can be
+   * overridden and used as a hook to perform any desired logic before passing
+   * the response to the validator.
+   * @param {{bytes: !ArrayBuffer, headers: !Headers}} response
+   * @protected
+   */
+  handleAdResponse(response) {
+    if (this.isStale_()) {
+      return;
+    }
+    const unvalidatedBytes = response.bytes;
+    const headers = response.headers;
+    if (!unvalidatedBytes) {
+      // TODO(levitzky) Add error reporting.
+      this.forceCollapse();
+      return;
+    }
+    dev().assert(this.boundValidator_, 'Validator never bound!');
+    this.unvalidatedBytes_ = unvalidatedBytes;
+    this.boundValidator_(unvalidatedBytes, headers, this)
+        .then(validatedBytes =>
+          this.handleValidatorResponse(validatedBytes));
+  }
+
+  handleAdResponseError(error) {
+    // TODO(levitzky) add actual error processing logic.
+    dev().warn(TAG, error);
+  }
+
+  /**
+   * Handler invoked when the current freshness id has expired.
+   */
+  handleStaleExecution() {}
+
+  /**
+   * Processes validator response and delegates further action to appropriate
+   *   renderer.
+   * @param {!ValidatorOutputDef} validatedResponse The utf-8 decoded ad
+   *   response.
+   */
+  handleValidatorResponse(validatedResponse) {
+    if (this.isStale_()) {
+      return;
+    }
+    if (!validatedResponse.creative) {
+      // TODO(levitzky) Add error reporting.
+      this.forceCollapse();
+      return;
+    }
+    dev().assert(this.boundRenderers_[validatedResponse.result],
+        'Renderer for AMP creatives never bound!');
+    this.boundRenderers_[validatedResponse.result](
+        this.getRenderingDataInput_(validatedResponse), this);
+  }
+
+
+  resetInstance() {
```

Comment:
what is this for?

Reply:
This is similar to [destroyFrame](https://github.com/ampproject/amphtml/blob/master/extensions/amp-a4a/0.1/amp-a4a.js#L1105) in amp-a4a. This would be called in unlayoutCallback.